### PR TITLE
Allow only one concurrent subset job

### PR DESCRIPTION
**Pull request by anguss00**

### CODE REVIEW

File: grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy [hunk 1]
```diff
@@ -21,13 +23,29 @@ class SubsetController {
         response.setContentType("application/octet-stream")
         response.setHeader("Content-disposition", "filename=${filenameToServe(subsetRequest)}")
 
-        try {
-            subsetService.subset(subsetRequest.typeName, subsetRequest.CQL_FILTER, response.outputStream)
+        if (isThreadAvailable()) {
```
Comment (julian1) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:26]: Is there a race condition between the testing of the mutex in isThreadAvailable(), and acquiring it in getThreadLock()? Eg. two or more threads could pass through isThreadAvailable() before getting to the next statement.

File: grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy [hunk 2]
```diff
@@ -21,13 +23,29 @@ class SubsetController {
...
+                getThreadLock()
+                subsetService.subset(subsetRequest.typeName, subsetRequest.CQL_FILTER, response.outputStream)
+            }
+            catch (InterruptedException e) {
+                log.info "The maximum number of concurrent jobs has been exceeded", e
+                render status: 503, text: "The maximum number of concurrent jobs has been exceeded, please try again later"
+            }
+            catch (Exception e) {
+                log.error "Unhandled exception during subset", e
+                releaseThreadLock()
```
Comment (julian1) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:37]: Is it necessary to release in this catch handler, if it's released in the finally block?

File: grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy (hunk 1 above)
Comment (danfruehauf) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:26]: This is not necessarily thread safe and can result in undeterministic behaviour.

What you want to have is basically 2 functions `up()` and `down()` - as you do, you can keep them in the names you're using now, it's fine. Eliminate the use of `isThreadAvailable()` and play only with those two. See if `acquire()` can take any parameters so you can specify when it actually fails. You want it to fail if there are too many locks.

Comment (danfruehauf) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:26]: (I wrote the last comment independently of what @julian1 wrote - but we saw the same thing).

Comment (anguss00) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:26]: The only reason I had to use `isThreadAvailable()` is that without this if statement (when using just `up()` and `down()`) any successive subset requests to an existing job will not trigger an interrupt exception and will continue to hang around in a queue in the background until a new thread is available for them. This is not behaviour we really want (as a job that takes 3 days could block up subsetter and after that's finished, the queued jobs just run automatically afterwards). I needed behaviour here such that if a job can't get a permit from the semaphore, it just dies then and there (hence if statement). But I agree with what you say above, that there is a race condition there that could cause trouble.

File: grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy (hunk 2 above)
Comment (anguss00) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:37]: Yes, we can definitely remove the release there, being over cautious haha

File: grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy (hunk 1 above)
Comment (anguss00) [inline grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy:26]: The issue that i have is that successive subset operations after the first don't trigger an interrupt on the running thread, instead they just queue up...

**Pull request closed without merging**